File: graph/traversal/dfs.py
import sys

# Fast I/O
input = sys.stdin.readline
sys.setrecursionlimit(200000)

class DFS:
    """
    DFS Template for CP
    Supports:
    - Recursive DFS
    - Iterative DFS
    - Entry/Exit times
    - Connected Components
    """
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.visited = [False] * (n + 1)
        self.entry = [-1] * (n + 1)
        self.exit = [-1] * (n + 1)
        self.timer = 0
    
    def dfs_recursive(self, u):
        """Standard recursive DFS with entry/exit times"""
        self.visited[u] = True
        self.entry[u] = self.timer
        self.timer += 1
        
        for v in self.adj[u]:
            if not self.visited[v]:
                self.dfs_recursive(v)
        
        self.exit[u] = self.timer
        self.timer += 1
    
    def dfs_iterative(self, start_node):
        """Iterative DFS using stack (avoids recursion limit)"""
        stack = [start_node]
        
        while stack:
            u = stack.pop()
            if not self.visited[u]:
                self.visited[u] = True
                # Process node u
                
                # Add neighbors to stack
                for v in reversed(self.adj[u]):
                    if not self.visited[v]:
                        stack.append(v)
    
    def solve(self):
        """Find connected components"""
        components = 0
        for i in range(1, self.n + 1):
            if not self.visited[i]:
                components += 1
                self.dfs_recursive(i)
        return components

def solve():
    # Example input: N nodes, M edges
    # 5 4
    # 1 2
    # 2 3
    # 1 4
    # 4 5
    
    try:
        line1 = input().split()
        if not line1: return
        n, m = map(int, line1)
        
        adj = [[] for _ in range(n + 1)]
        for _ in range(m):
            u, v = map(int, input().split())
            adj[u].append(v)
            adj[v].append(u)  # Undirected
            
        dfs_solver = DFS(n, adj)
        cc = dfs_solver.solve()
        print(f"Connected Components: {cc}")
        
    except ValueError:
        pass

if __name__ == "__main__":
    solve()

File: graph/traversal/dfs.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * DFS Template for Competitive Programming
 * Time Complexity: O(V + E)
 * Space Complexity: O(V)
 */

const int MAXN = 200005;
vector<int> adj[MAXN];
bool visited[MAXN];
int entry[MAXN], leave[MAXN];
int timer;

void dfs(int u) {
    visited[u] = true;
    entry[u] = timer++;
    
    for (int v : adj[u]) {
        if (!visited[v]) {
            dfs(v);
        }
    }
    
    leave[u] = timer++;
}

void dfs_iterative(int start_node) {
    stack<int> s;
    s.push(start_node);
    
    while (!s.empty()) {
        int u = s.top();
        s.pop();
        
        if (!visited[u]) {
            visited[u] = true;
            // Process node u
            
            for (auto it = adj[u].rbegin(); it != adj[u].rend(); ++it) {
                if (!visited[*it]) {
                    s.push(*it);
                }
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;
    
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u); // Undirected
    }
    
    // Reset
    fill(visited, visited + n + 1, false);
    timer = 0;
    
    int components = 0;
    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            components++;
            dfs(i);
        }
    }
    
    cout << "Connected Components: " << components << "\n";
    
    return 0;
}
